[Feat][TNA] Fabric Components Intro (#3132)
facebook/react-native-website: docs/the-new-architecture/pillars-fabric-components.md

title: Fabric Components
---

> A Fabric Component is a UI component which is rendered on the screen using the [Fabric Renderer](https://reactnative.dev/architecture/fabric-renderer).

Reviewer: ```suggestion
A Fabric Component is a UI component rendered on the screen using the [Fabric Renderer](https://reactnative.dev/architecture/fabric-renderer).
```

Just more concise.

> Using Fabric Components instead of Native Components allows us to reap all the [benefit](./why) coming from the **New Architecture**. Specifically, we are able to leverage JSI to efficiently connect the Native UI code and the JavaScript one, and all the other features provided by the new renderer.

Reviewer: ```suggestion
Using Fabric Components instead of Native Components allows us to reap all the [benefits](./why) of the **New Architecture**. Specifically, we are able to leverage JSI to efficiently connect the Native UI and JavaScript.
```

Explanations:

* "the Native UI code and the JavaScript one" -- "code" and "one" don't match in this phrase, because "code" is [an uncountable noun](https://dictionary.cambridge.org/us/grammar/british-grammar/nouns-countable-and-uncountable). Using a countable noun like "layer" ("the Native UI layer and the JavaScript one") would be more correct, but it still sounds off to me in this phrase. "the Native UI code and the JavaScript code" would be better, as would the shorter phrase I suggested above.
* I'm wondering what the second sentence is attempting to teach the reader. As it stands, I'm not sure it adds useful information or makes the document flow better.
  * Is there something more specific we can say that "efficiently connect"? To the reader, this does not give me any new specific understanding. If not, I'd opt for removing it.
  * "all the other features provided" also does not help me understand anything new, so I'd either make it more specific or remove it entirely.
  * To make this sentence more "useful", it could be combined with the next paragraph, where we explain each step and why it's useful together (e.g., "A Fabric Component starts with a JS specification, which Codegen uses to generate some C++ code for use on the native side. This helps ensure your JS and Native interfaces have matching specifications" (first draft, I think that's possible to write better)

Author: Thanks for the feedback, I applied the suggestion. 

About this:
> To make this sentence more "useful", it could be combined with the next paragraph, where we explain each step and why it's useful together (e.g., "A Fabric Component starts with a JS specification, which Codegen uses to generate some C++ code for use on the native side. This helps ensure your JS and Native interfaces have matching specifications" (first draft, I think that's possible to write better)

Do you mean that we should delete the "Using Fabric Components ..." sentence or just put the two paragraph together with the suggested rewording?

> A Fabric Component is created starting from a **JS specification**. This, with the help of a bit of [**Codegen**](./pillars-codegen), will create some C++ code, shared among all the RN's platforms. Then, after filling in the missing **native code** with Component specific logic, the Component can be **integrated** in the app.

Reviewer: ```suggestion
A Fabric Component is created starting from a **JS specification**. This, with the help of [**Codegen**](./pillars-codegen), will create some C++ code, shared among all React Native platforms. Then, after filling in the missing **native code** with component-specific logic, the Component can be integrated in the app.
```

Explanations:
* "Codegen" is a specific tool, so "a bit of" doesn't quite work to my ear, we can just say "with the help of Codegen"
* "C++ code..." I might consider rephrasing this further to specifically say that the C++ is integrated with the platform-native code, it might be unclear to a new reader
* "component-specific logic" needs to be hyphenated, because "component-specific" is acting as [a compound adjective on the noun "logic"](https://www.grammarly.com/blog/hyphen/)
* "integrated" is not used as a new or special term here, I wouldn't bold it ("native code" is similar, but closer to a "special term" the way RN docs treat it)
* "filling in the missing native code" is an ambiguous phrase to me. What is missing, from where? If we're talking about just writing the native code to bridge between C++ and JS, I would phrase it as "adding code that bridges..." or "connecting", instead of "filling in missing", because "missing" sounds like something was supposed to be there but something went wrong.

Author: I tried to rephrase this bit. Let me know what do you think!

The following section will guide you through the creation of a Fabric Component, step-by-step.

:::caution
Fabric Components only works with the **New Architecture** enabled.
To migrate to the **New Architecture**, follow the [Migration guide](../new-architecture-intro)
:::

## How to Create a Fabric Components

To create a Fabric Component, we have to follow these steps:

> - Define a set of JS specification.

Author: > Style question: we're referring to JS/JavaScript/Javascript a few different ways. Should we standardize on "JavaScript"? (I don't mind occasionally abbreviating it when brevity is useful)

When we have to spell it out completely, it has to be JavaScript (and TypeScript, accordingly). I'm still trying to figure out when it makes sense to shorten them.. I'm starting leaning toward not shortening them at all... 😅

> - Configure the component so that it can be consumed by an app.

Author: When I specify that is a Fabric Component, I use capital letters (it's a proper name for an Entity in the architecture). When I refer to the entity as a generic item belonging to that class, I used lowercase letter, also not to clutter too much the article.
Does it make sense? Do we prefer to have always a capital C for component?

> - Configure the component so that it can be consumed by an app.

Reviewer 2: Ok makes sense to me 👍 I'm a big fan of title case, in a sense that if you write `Configure the Component so that` you're implicitely referring to the Fabric Component.

It also makes for more consistent reading for me, but that's really a nit :) up to your preference

> - Configure the component so that it can be consumed by an app.

Author: :+1: I updated the PR. The only lowercase `component` remained is in the first line, which is used together with `UI` so it refers to the class of `UI` components rather than to `Fabric Components`

> - Configure the component so that it can be consumed by an app.

Reviewer 2: Awesome 👌 resolving here.

> - Write the native code required to make it works.

Author: Thanks, I alway makes mistakes with the `make it` construct! 😅

The following are the specification of our `RTNCenteredText` component in both Flow and TypeScript: let's create a `RTNCenteredText` file with the proper extension in the `js` folder.

<Tabs groupId="fabric-component-specs" defaultValue={constants.defaultJavaScriptSpecLanguages} values={constants.javaScriptSpecLanguages}>
<TabItem label='flow'>

```typescript
// @flow strict-local

import type {ViewProps} from 'react-native/Libraries/Components/View/ViewPropTypes';
import type {HostComponent} from 'react-native';
import codegenNativeComponent from 'react-native/Libraries/Utilities/codegenNativeComponent';

type NativeProps = $ReadOnly<{|
  ...ViewProps,
  text: ?string,
  // add other props here
|}>;

export default (codegenNativeComponent<NativeProps>(
   'RTNCenteredText',
): HostComponent<NativeProps>);
```

</TabItem>
<TabItem label='typescript'>

```typescript

> import type { ViewProps } from 'ViewPropTypes';

Author: Discussion started here: https://github.com/facebook/react-native-website/pull/3133#discussion_r898345765

@lindboe

import type { HostComponent } from 'react-native';
import codegenNativeComponent from 'react-native/Libraries/Utilities/codegenNativeComponent';

export interface NativeProps extends ViewProps {
  ...ViewProps,
  text: string | null | undefined,
  // add other props here
}

export default codegenNativeComponent<NativeProps>(
  'RTNCenteredText'

> ) as HostComponent<NativeProps>;

Author: Discussion started here: https://github.com/facebook/react-native-website/pull/3133#discussion_r898346350

@lindboe

```

</TabItem>
</Tabs>

Let's break these down a little.

At the beginning of the spec files, there are the imports. Here, we can import what we need from React Native and other dependencies if needed. The most important imports, that are present in all the Fabric Components, are:

> - The `HostComponent`, which is the type our exported component needs to conform to;

Author: Discussion started here: https://github.com/facebook/react-native-website/pull/3133#discussion_r898327002

Author: @lindboe

```json
{
  "name": "rnt-centered-text",
  "version": "0.0.1",
  "description": "Showcase a Fabric Component with a centered text",
  "react-native": "js/index",
  "source": "js/index",
  "files": [
    "js",
    "android",
    "ios",
    "rnt-centered-text.podspec",
    "!android/build",
    "!ios/build",
    "!**/__tests__",
    "!**/__fixtures__",
    "!**/__mocks__"
  ],
  "keywords": ["react-native", "ios", "android"],
  "repository": "https://github.com/<your_github_handle>/rnt-centered-text",
  "author": "<Your Name> <your_email@your_provider.com> (https://github.com/<your_github_handle>)",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/<your_github_handle>/rnt-centered-text/issues"
  },
  "homepage": "https://github.com/<your_github_handle>/rnt-centered-text#readme",
  "devDependencies": {},
  "peerDependencies": {
    "react": "*",
    "react-native": "*"
  },
  "codegenConfig": {
    "libraries": [
      {
        "name": "RTNCenteredTextSpecs",
        "type": "components",
        "jsSrcsDir": "js"
      }
    ]
  }
}

> ```

Author: Discussion started here: https://github.com/facebook/react-native-website/pull/3136#discussion_r899844605

@Simek

- Write the native code that will make it work

When developing a React Native app that uses a Fabric Component, it is responsibility of the app to actually generate the code using **CodeGen**. However, when developing a Fabric Component as a library, we need to reference the generated code and it is therefore useful to see what the app will generate.

As first step for both iOS and Android, this guide shows how to execute manually the scripts used by **CodeGen** to generate the required code. Further information on **CodeGen** can be found [here](/docs/pillars-codegen.md)

:::caution
The code generated by the **CodeGen** in this step should not be committed to the versioning system. React Native apps are able to generate the code when the app is built. This allows to avoid any ABI incompatibility and to ensure that a consistent version of the **CodeGen** is used.
:::

### iOS

#### Generate the code - iOS

To run Codegen for the iOS platform, we need to open a terminal and run the following command:

```sh
cd MyApp
yarn add ../RTNCenteredText
cd ..
node MyApp/node_modules/react-native/scripts/generate-artifacts.js \
  --path MyApp/ \

>   --outputPath RTNCenteredText/generated/

Author: Discussion started here: https://github.com/facebook/react-native-website/pull/3148#discussion_r898688072

@lindboe

                        │   ├── EventEmitters.cpp
                        │   ├── EventEmitters.h
                        │   ├── Props.cpp
                        │   ├── Props.h
                        │   ├── RCTComponentViewHelpers.h
                        │   ├── ShadowNodes.cpp
                        │   └── ShadowNodes.h
                        └── rncore
                            ├── ComponentDescriptors.h
                            ├── EventEmitters.cpp
                            ├── EventEmitters.h
                            ├── Props.cpp
                            ├── Props.h
                            ├── RCTComponentViewHelpers.h
                            ├── ShadowNodes.cpp
                            └── ShadowNodes.h
```

The relevant path for the Component we are writing is `generated/build/generated/ios/react/renderer/components/RTNCenteredTextSpecs`.
This folder contains all the generated code required by our Component.

See the [CodeGen](./pillars-codegen) section for further details on the generated files.

#### Write the Native iOS Code

Now that we can see the iOS code we need, it's time to write the Native code for our Fabric Component.

> We need to create three files in the `RTNCenteredText/ios` folder:

Author: Discussion started here: https://github.com/facebook/react-native-website/pull/3148#discussion_r898691780

@lindboe

title: Fabric Components
---

import Tabs from '@theme/Tabs'; import TabItem from '@theme/TabItem'; import constants from '@site/core/TabsConstants';

A Fabric Component is a UI component rendered on the screen using the [Fabric Renderer](https://reactnative.dev/architecture/fabric-renderer).

Using Fabric Components instead of Native Components allows us to reap all the [benefits](./why) of the **New Architecture**. Specifically, it enables to use JSI to efficiently connect the Native UI code with JavaScript.

> A Fabric Component is created starting from a **JavaScript specification**. Then [**CodeGen**](./pillars-codegen) creates some C++ code, integrated in the platform native layer and shared among all the React Native platforms. This C++ code is boilerplate code that component-specific logic needs to import and use to be properly adopted by React Native. After the component-specific logic is connected with the generated code, the component can be integrated in the app.

Reviewer: This rephrasing looks pretty good! I have just a few smaller questions:

* "integrated in the platform native layer and shared among all the React Native platforms": I also took a stab at rephrasing a similar idea in the TM guide, let me know what you think: https://github.com/facebook/react-native-website/pull/3168/files#diff-ba6eb9aa37edbf22f752c4267cb34a52ff888f9a12d1795335fdfaf6dfc61f05R13.
  * I am thinking "integrated in the platform native layer" and "shared among all the RN platforms" might sound like the same thing to some readers (although I know what you mean) -- they might both sound like "C++ code that the native language you're used to seeing on each platform can use"). I think to disambiguate the two ideas here we might want a more specific phrasing than "integrate".
* I think "that component-specific logic needs to import and use to be properly adopted by React Native" still has some ambiguity for the reader. What is "component-specific logic"? What does "properly adopted" mean? Brainstorming a few ideas for other phrasing:
  * "This C++ code is scaffolding that your native component logic needs to utilize to work with the New Architecture"
  * "Then Codegen creates some C++ code, generating logic and interfaces that will help bridge your JavaScript code to the Native Platform. You can either add more C++ code or integrate the generated code with your Native Platform language of choice." (Also replaces the sentence before).
* nit: "boilerplate code": this is perfectly fine, but I really liked @cortinico's suggestion of "scaffolding code" and used it a few times in the TM guide, so we might want to consider being consistent about that

Author: Thanks Lizzi! I used your list to give consistency between the two guides. I then rewrote the paragraph on the Codegen, simplifying it a little bit and adding some examples to better explain the idea.

Then there are the dependencies for this package, specifically you need `react` and `react-native`, and you can add all the other dependencies you may have.

Finally, the **CodeGen** configuration is specified by the `codegenConfig` field. It contains an array of libraries, each of which is defined by three other fields:

- `name`: The name of the library. By convention, you should add the `Specs` suffix.
- `type`: The type of module contained by this package. In this case, it is a component, thus the value to use is `components`.
- `jsSrcsDir`: the relative path to access the `js` specification that is parsed by the **CodeGen**.

### iOS: Create the `podspec` file

Now, let's configure the native component to generate the required code to integrate it with an app. For further information on how the **CodeGen**, have a look [here](./pillars-codegen.md).

For iOS, you need to create a `.podspec` file which defines the component as a dependency. This file stays in the root folder of the `RTNCenteredText` component, _outside_ the `ios` folder.

> The `.podspec` file for the component looks like this

Author: I took your section 3 and ported to the Fabric Guide. Thanks for the rewriting!

    url "$projectDir/../node_modules/react-native/android"
  }
  mavenCentral()
  google()
}

dependencies {
  implementation 'com.facebook.react:react-native:+'
}

react {
    jsRootDir = file("../js/")
    libraryName = "RTNCenteredText"
    codegenJavaPackageName = "com.RTNCenteredText"
}
```

The `build.gradle` file contains various section to properly configure the library and to access the required dependencies. The most interesting sections are:

- the `defaultConfig` block within the `android` block, where a `buildConfigField` is added to enable the new architecture.
- the `react` block where you configure the **CodeGen** process. Here, the following properties are required:
  - The `jsRootDir` contains the relative path to the JavaScript specs.
  - The `libraryName` is used to link the library to the app.
  - The `codegenJavaPackageName` that corresponds to the name of the Java Package that is used to group the code generated by **CodeGen**.

## 4. Native Code

The last step requires you to write some native code to connect the JavaScript side of the Component to what is offered by the platforms. This process requires two main steps:

> 1. Run the **CodeGen** to see what would be generated.

Author: I have updated all the _"the **Codegen"_

:::note
To run the **CodeGen**, you need to enable the **New Architecture** in the Android app. This can be done by opening the `gradle.properties` files and by switching the `newArchEnabled` property from `false` to `true`.
:::

The generated code is stored in the `MyApp/node_modules/rnt-centered-text/android/build/generated/source/codegen` folder and it has this structure:

```title="Android generated code"
codegen
├── java
│   └── com
│       └── facebook
│           └── react
│               └── viewmanagers
│                   ├── RTNCenteredTextManagerDelegate.java
│                   └── RTNCenteredTextManagerInterface.java
├── jni
│   ├── Android.mk
│   ├── CMakeLists.txt
│   ├── RTNCenteredText-generated.cpp
│   ├── RTNCenteredText.h
│   └── react
│       └── renderer
│           └── components
│               └── RTNCenteredText
│                   ├── ComponentDescriptors.h
│                   ├── EventEmitters.cpp
│                   ├── EventEmitters.h
│                   ├── Props.cpp
│                   ├── Props.h
│                   ├── ShadowNodes.cpp
│                   └── ShadowNodes.h
└── schema.json
```

> You can see that the content of the `codegen/jni/react/renderer/components/RTNCenteredTextSpecs` looks similar to the files created by the iOS counterpart. Other interesting pieces are the `Android.mk` and `CMakeList.txt` files, which you need to configure the Fabric Component in the app. And the `RTNCenteredTextManagerDelegate.java` and `RTNCenteredTextManagerInterface.java` that you need to use in our manager.

Author: No, good point. I'll try to add a short paragraph for that!

- The use of JSI, a JavaScript interface for native code, which allows for more efficient communication between native and JavaScript code than the bridge.

A Fabric Component is created starting from a **JavaScript specification**. Then [**Codegen**](./pillars-codegen) creates some C++ scaffolding code to connect the component-specific logic (for example, accessing some native platform capability) to the rest of the React Native infrastructure. The C++ code is the same for all the platforms. Once the component is properly connected with the scaffolding code, it is ready to be imported and used by an app.

The following section guides you through the creation of a Fabric Component, step-by-step.

:::caution
Fabric Components only works with the **New Architecture** enabled.
To migrate to the **New Architecture**, follow the [Migration guide](../new-architecture-intro)
:::

## How to Create a Fabric Components

To create a Fabric Component, you have to follow these steps:

1. Define a set of JavaScript specifications.

> 2. Configure the component so that **Codegen** can create the shared code and it can added as a dependency for an app.

Reviewer 2: ```suggestion
2. Configure the component so that **Codegen** can create the shared code and it can be added as a dependency for an app.
```

> 3. Write the required native code required.

Reviewer 2: ```suggestion
3. Write the required native code.
```

        return new RTNCenteredText(context);
    }

    @Override
    @ReactProp(name = "text")
    public void setText(RTNCenteredText view, @Nullable String text) {
        view.setText(text);
    }
}
```

The `RTNCenteredTextManager` is a class used by React Native to instantiate the native component. It is the class that implements the interfaces generated by **Codegen** (see the `RTNCenteredTextManagerInterface` interface in the `implements` clause) and it uses the `RTNCenteredTextManagerDelegate` class.

It is also responsible for exporting all the constructs required by React Native: the class itself is annotated with `@ReactModule` and the `setText` method is annotated with `@ReactProp`.

##### RTNCenteredTextPackage.java

Finally, open the `RTNCenteredTextPackage.java` file in the `android/src/main/java/com/rtncenteredtext` folder and update it with the following lines

```diff title="RTNCenteredTextPackage update"
package com.rtncenteredtext;

import com.facebook.react.ReactPackage;
import com.facebook.react.bridge.NativeModule;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.uimanager.ViewManager;

> + import java.util.ArrayList;

Reviewer 2: This is not needed then

import java.util.Collections;
import java.util.List;

public class RTNCenteredTextPackage implements ReactPackage {

    @Override
    public List<ViewManager> createViewManagers(ReactApplicationContext reactContext) {
-        return Collections.emptyList();

> +        List<ViewManager> viewManagers = new ArrayList<>();
> +        viewManagers.add(new RTNCenteredTextManager(reactContext));
> +        return viewManagers;

Reviewer 2: ```suggestion
+        return Collections.singletonList(new RTNCenteredTextManager(reactContext));
```

This can be simplified this way
